kart/spatial_tree.py

```diff
@@ -0,0 +1,413 @@
+        ds_path = m.group(1)
+        yield ds_path, oid
+
+
+class CrsHelper:
+    """
+    Loads all CRS definitions for a particular dataset,
+    and creates transforms
+    """
+
+    def __init__(self, repo):
+        self.repo = repo
+        self.ds_to_transforms = {}
+
+    def transforms_for_dataset(self, ds_path):
+        transforms = self.ds_to_transforms.get(ds_path)
+        if transforms is None:
+            transforms = self._load_transforms_for_dataset(ds_path)
+            self.ds_to_transforms[ds_path] = transforms
+        return transforms
+
+    def _load_transforms_for_dataset(self, ds_path):
+        if ds_path in self.ds_to_transforms:
+            return self.ds_to_transforms[ds_path]
+
+        crs_oids = set(self.iter_crs_oids(ds_path))
+        transforms = []
+        descs = []
+        for crs_oid in crs_oids:
+            try:
+                transform, desc = self.transform_from_oid(crs_oid)
+                if transform not in transforms:
+                    transforms.append(transform)
+                    descs.append(desc)
+            except Exception as e:
+                L.warning(
+                    f"Couldn't load transform for CRS {crs_oid} at {ds_path}\n{e}"
+                )
+        L.info(f"Loaded CRS transforms for {ds_path}: {', '.join(descs)}")
+        return transforms
+
+    def iter_crs_oids(self, ds_path):
+        cmd = (
+            _revlist_command(self.repo)
+            + ["--all", "--"]
+            + list(self.all_crs_paths(ds_path))
+        )
+        try:
+            r = subprocess.run(
+                cmd,
+                encoding="utf8",
+                check=True,
+                capture_output=True,
+                env=tool_environment(),
+            )
+        except subprocess.CalledProcessError as e:
+            raise SubprocessError(
+                f"There was a problem with git rev-list: {e}", called_process_error=e
+            )
+        for d, crs_oid in _parse_revlist_output(
+            r.stdout.splitlines(), r"meta/crs/[^/]+"
+        ):
+            assert d == ds_path
+            yield crs_oid
+
+    def all_crs_paths(self, ds_path):
+        # Delete .sno-dataset if we drop V2 support.
+        yield f"{ds_path}/.sno-dataset/meta/crs/"
+        yield f"{ds_path}/.table-dataset/meta/crs/"
+
+    @functools.lru_cache()
+    def transform_from_oid(self, crs_oid):
+        wkt = normalise_wkt(self.repo[crs_oid].data.decode("utf-8"))
+        return self.transform_from_wkt(wkt)
+
+    @functools.lru_cache()
+    def transform_from_wkt(self, wkt):
+        src_crs = make_crs(wkt)
+        if src_crs.IsGeographic():
+            transform = None
+            desc = f"IDENTITY({src_crs.GetAuthorityCode(None)})"
+        else:
+            target_crs = src_crs.CloneGeogCS()
+            transform = osr.CoordinateTransformation(src_crs, target_crs)
+            desc = f"{src_crs.GetAuthorityCode(None)} -> {target_crs.GetAuthorityCode(None)}"
```

Comment:
this is incorrect I think - we don't want to project to the nearest geographic CRS, we want to project everything to a single CRS. Otherwise S2 cells won't be comparable between different source CRSes.

Different geographic CRSes have different meridians, axis order, possibly units (e.g. radians) etc. We should pick one and always transform to that. Normal choice being EPSG:4326

Reply:
Hmm I think you are right. However again I must defer to @rcoup who wrote the prototype - I imagine he had his reasons for doing it this way. If not, I'll change it.

The "Geocentric vs Geodetic coordinates" section in https://s2geometry.io/about/overview might be relevant, but I'm not even sure about that.

Reply:
Logic was avoiding needless & expensive datum transforms. But Craig's right, axis-order/units/meridians can be in different places so that's not a great plan.

WGS84 seems sensible to me.

Reply:
Okay. Will switch this to convert to WGS 84. Is there a way we can tell it not to bother doing the expensive and needless thing?

Reply:
`OSRSpatialReference.IsSame(other)`?